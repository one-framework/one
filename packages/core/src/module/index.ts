export * from './scanner';
export * from './module';
export * from './compiler';
export * from './container';
export * from './injection-token';
