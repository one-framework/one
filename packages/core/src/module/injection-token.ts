export class InjectionToken<T> {
  public readonly context = `InjectionToken<${this.desc}>`;
  public readonly name = Symbol.for(this.context);

  constructor(private readonly desc: string) {}
}
