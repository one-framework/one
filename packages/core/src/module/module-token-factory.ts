import * as hash from 'object-hash';
import stringify from 'fast-safe-stringify';

import { Type, DynamicModule } from '../interfaces';
import { Reflector } from '../reflector';

export class ModuleTokenFactory {
  public create(
    target: Type<any>,
    scope: Type<any>[],
    dynamicModuleMetadata?: Partial<DynamicModule>,
  ): string {
    const reflectedScope = Reflector.getModuleScope(target);
    const isSingleScoped = reflectedScope === true;

    const opaqueToken = {
      module: target.name,
      dynamic: this.getDynamicMetadataToken(dynamicModuleMetadata!),
      scope: isSingleScoped ? this.getScopeStack(scope) : reflectedScope,
    };

    return hash(opaqueToken);
  }

  private getDynamicMetadataToken(
    dynamicModuleMetadata: Partial<DynamicModule>,
  ) {
    // Uses safeStringify instead of JSON.stringify
    // to support circular dynamic modules
    return dynamicModuleMetadata ? stringify(dynamicModuleMetadata) : '';
  }

  public getScopeStack(scope: Type<any>[]): string[] {
    const reversedScope = scope.reverse();
    const firstGlobalIndex = reversedScope.findIndex(
      s => Reflector.getModuleScope(s) === 'global',
    );
    scope.reverse();

    const stack =
      firstGlobalIndex >= 0
        ? scope.slice(scope.length - firstGlobalIndex - 1)
        : scope;
    return stack.map(module => module.name);
  }
}
