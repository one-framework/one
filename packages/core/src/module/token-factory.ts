import hash from 'object-hash';
import stringify from 'fast-safe-stringify';

import { Type, ModuleMetadata } from '../interfaces';
import { Reflector } from '../reflector';

export class ModuleTokenFactory {
  public create(
    target: Type,
    scope: Type[],
    dynamicModuleMetadata?: ModuleMetadata,
  ): string {
    const reflectedScope = Reflector.getModuleScope(target);
    const isSingleScoped = reflectedScope === true;

    const opaqueToken = {
      module: target.name,
      dynamic: this.getDynamicMetadataToken(dynamicModuleMetadata!),
      scope: isSingleScoped ? this.getScopeStack(scope) : reflectedScope,
    };

    return hash(opaqueToken);
  }

  private getDynamicMetadataToken(
    dynamicModuleMetadata: ModuleMetadata,
  ): string {
    // Uses safeStringify instead of JSON.stringify
    // to support circular dynamic modules
    return dynamicModuleMetadata ? stringify(dynamicModuleMetadata) : '';
  }

  private getScopeStack(scope: Type[]): string[] {
    const reversedScope = scope.reverse();
    const firstGlobalIndex = reversedScope.findIndex(
      s => Reflector.getModuleScope(s) === 'global',
    );
    scope.reverse();

    const stack =
      firstGlobalIndex >= 0
        ? scope.slice(scope.length - firstGlobalIndex - 1)
        : scope;
    return stack.map(module => module.name);
  }
}
