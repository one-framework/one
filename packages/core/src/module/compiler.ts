import {
  DynamicModule,
  Type,
  ModuleFactory,
  ModuleImport,
  Omit,
} from '../interfaces';
import { ModuleTokenFactory } from './token-factory';
import { Registry } from '../registry';
import { OneModule } from './module';
import { getDeferred } from '../util';

export class ModuleCompiler {
  private readonly moduleTokenFactory = new ModuleTokenFactory();

  public async compile(
    module: ModuleImport,
    scope: Type<OneModule>[] = [],
  ): Promise<ModuleFactory> {
    const { target, dynamicMetadata } = await this.extractMetadata(module);

    const token = this.moduleTokenFactory.create(
      target,
      scope,
      dynamicMetadata,
    );

    return { target, dynamicMetadata, token };
  }

  private async extractMetadata(
    module: ModuleImport,
  ): Promise<Omit<ModuleFactory, 'token'>> {
    const moduleRef = await getDeferred(module);

    if (!Registry.isDynamicModule(moduleRef)) {
      return { target: <Type>module };
    }

    const { module: target, ...dynamicMetadata } = <DynamicModule>moduleRef;
    return { target, dynamicMetadata };
  }
}
