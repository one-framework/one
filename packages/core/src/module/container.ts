import { InjectionToken } from './injection-token';
import { ModuleCompiler } from './compiler';
import { Reflector } from '../reflector';
import { Registry } from '../registry';
import { OneModule } from './module';
import { Metadata } from '../constants';
import { isNil, flatten, concat, getValues } from '../util';
import {
  UnknownModuleException,
  InvalidModuleException,
  UnknownProviderException,
  MissingInjectionTokenException,
} from '../errors';
import {
  Dependency,
  DynamicModule,
  ModuleExport,
  ModuleImport,
  Provider,
  Token,
  Type,
} from '../interfaces';

export interface StrictSelect {
  strict?: boolean;
}

export class OneContainer {
  private readonly moduleCompiler = new ModuleCompiler();
  private readonly globalModules = new Set<OneModule>();
  private readonly modules = new Map<string, OneModule>();
  public readonly moduleOrder = new Set<OneModule>();
  public readonly providerTokens: Token[] = [];
  private readonly dynamicModulesMetadata = new Map<
    string,
    Partial<DynamicModule>
  >();

  private getModulesFrom(module?: Type<OneModule>) {
    return !isNil(module)
      ? [<OneModule>this.getModule(module)]
      : this.getModuleValues();
  }

  public isProviderBound(
    provider: Type<any> | InjectionToken<any>,
    module?: Type<OneModule>,
  ) {
    const token = Registry.getProviderToken(provider);
    return this.getModulesFrom(module).some(({ providers }) =>
      providers.isBound(token),
    );
  }

  public replace(
    toReplace: Dependency,
    options: any & { scope: any[] | null },
  ) {
    [...this.modules.values()].forEach(module => {
      module.replace(toReplace, options);
    });
  }

  public getProvider<T>(
    provider: Token | Type<T> | InjectionToken<T>,
    scope?: Type<OneModule>,
    { strict }: StrictSelect = {},
  ): T {
    const token = Registry.getProviderToken(provider);

    if (strict) {
      const module = this.getModule(scope!);
      if (module!.providers.isBound(token)) {
        return module!.providers.get(token);
      }
    } else {
      for (const { providers } of this.modules.values()) {
        if (providers.isBound(token)) {
          return providers.get<T>(token);
        }
      }
    }

    throw new UnknownProviderException(<any>provider, scope!);
  }

  public getAllProviders<T>(
    provider: InjectionToken<T>,
    target?: Type<OneModule>,
  ) {
    if (!Registry.isInjectionToken(provider)) {
      throw new MissingInjectionTokenException('Container.getAllProviders()');
    }

    const token = Registry.getProviderToken(provider);

    return flatten<T | Promise<Type<Provider>>>(
      this.getModulesFrom(target).map(({ providers }) =>
        providers.isBound(token) ? providers.getAll(token) : [],
      ),
    );
  }

  public getModuleValues() {
    return getValues<OneModule>(this.modules.entries());
  }

  public hasModule(module: Type<any>) {
    return this.getModuleValues().some(({ target }) => target === module);
  }

  public getModule(module: Type<any>): OneModule | undefined {
    return this.getModuleValues().find(({ target }) => target === module);
  }

  public getModuleByToken(token: string) {
    if (!this.modules.has(token)) {
      throw new UnknownModuleException([token]);
    }

    return <OneModule>this.modules.get(token);
  }

  public getReversedModules() {
    return [...this.modules.entries()].reverse();
  }

  public getModules() {
    return this.modules;
  }

  public async addProvider(provider: Provider, token: string) {
    const module = this.getModuleByToken(token);
    await module.addProvider(provider);
  }

  public addExported(component: ModuleExport, token: string) {
    const module = this.getModuleByToken(token);
    module.addExported(component);
  }

  public addGlobalModule(module: OneModule) {
    this.globalModules.add(module);
  }

  public async addModule(
    module: Partial<ModuleImport>,
    scope: Type<OneModule>[] = [],
  ) {
    if (!module) throw new InvalidModuleException(scope);

    const {
      target,
      dynamicMetadata,
      token,
    } = await this.moduleCompiler.compile(<any>module, scope);
    if (this.modules.has(token)) return;

    const oneModule = new OneModule(target, scope, this);
    oneModule.addGlobalProviders();
    this.modules.set(token, oneModule);

    const modules = concat(scope, target);
    this.addDynamicMetadata(token, dynamicMetadata!, modules);

    if (Reflector.isGlobalModule(target)) {
      this.addGlobalModule(oneModule);
    }
  }

  private addDynamicMetadata(
    token: string,
    dynamicModuleMetadata: Partial<DynamicModule>,
    scope: Type<OneModule>[],
  ) {
    if (!dynamicModuleMetadata) return;

    this.dynamicModulesMetadata.set(token, dynamicModuleMetadata);
    this.addDynamicModules(dynamicModuleMetadata.imports, scope);
  }

  private addDynamicModules(
    modules: ModuleImport[] = [],
    scope: Type<OneModule>[],
  ) {
    modules.forEach(module => this.addModule(module, scope));
  }

  public bindGlobalScope() {
    this.modules.forEach(module => this.bindGlobalsToImports(module));
  }

  private bindGlobalsToImports(module: OneModule) {
    this.globalModules.forEach(globalModule =>
      this.bindGlobalModuleToModule(module, globalModule),
    );
  }

  private bindGlobalModuleToModule(module: OneModule, globalModule: OneModule) {
    if (module === globalModule) return;
    module.addImport(globalModule);
  }

  public async addImport(relatedModule: Partial<ModuleImport>, token: string) {
    if (!this.modules.has(token)) return;

    const module = this.getModuleByToken(token);
    const scope = concat(module.scope, module.target);

    const { token: relatedModuleToken } = await this.moduleCompiler.compile(
      <any>relatedModule,
      scope,
    );

    const related = this.getModuleByToken(relatedModuleToken);
    module.addImport(related);
  }

  public getDynamicMetadataByToken(token: string, key: Metadata) {
    const metadata = this.dynamicModulesMetadata.get(token);
    return metadata && metadata[key] ? metadata[key] : [];
  }
}
