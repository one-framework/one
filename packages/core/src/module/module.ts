import getDecorators from 'inversify-inject-decorators';

import { ProviderTypes, Scopes } from '../constants';
import { OneContainer } from './container';
import { Reflector } from '../reflector';
import { Registry } from '../registry';
import { createDeferredPromise, runSeries } from '../util';
import { InjectionToken } from './injection-token';
import {
  InvalidExportException,
  MultiProviderException,
  UnknownExportException,
  UnknownProviderException,
} from '../errors';
import {
  ClassProvider,
  Dependency,
  ExistingProvider,
  FactoryProvider,
  ModuleExport,
  ModuleImport,
  MultiDepsProvider,
  Provider,
  Token,
  Type,
  ValueProvider,
} from '../interfaces';
import {
  APP_DESTROY,
  APP_INIT,
  Injector,
  MODULE_INIT,
  MODULE_REF,
} from '../tokens';

export class OneModule {
  public readonly imports = new Set<OneModule>();
  public readonly providers = new Set<Provider>();
  public readonly injector = new Injector({ skipBaseClassChecks: true });
  public readonly lazyInject = getDecorators(this.injector).lazyInject;
  public readonly exports = new Set<Token>();
  public readonly created = createDeferredPromise();
  private readonly factoryValues = new Map<Token, any>();

  constructor(
    public readonly target: Type,
    public readonly scope: Type[],
    public readonly container: OneContainer,
  ) {}

  public addImport(relatedModule: OneModule) {
    this.imports.add(relatedModule);
  }

  public addProvider(provider: Provider) {
    this.providers.add(provider);
  }

  private providerContainerHasToken(token: Token) {
    return [...this.providers.values()].some(provider => {
      return Registry.getProviderToken(provider) === token;
    });
  }

  private validateExported(token: Token, exported: ModuleExport) {
    if (this.providerContainerHasToken(token)) return token;

    const imported = [...this.imports.values()];
    const importedRefNames = <any[]>imported
      .filter(item => item)
      .map(({ target }) => target)
      .filter(target => target);

    if (!importedRefNames.includes(token)) {
      throw new UnknownExportException(this.target.name, (<Type>exported).name);
    }

    return token;
  }

  public addExported(exported: ModuleExport) {
    const addExportedUnit = (token: Token) => {
      this.exports.add(this.validateExported(token, exported));
    };

    if (Registry.isDynamicModule(exported)) {
      return addExportedUnit(exported.module);
    }

    addExportedUnit(Registry.getProviderToken(exported));
  }

  public getProviders() {
    return [
      Injector,
      OneContainer,
      MODULE_REF,
      ...this.providers.values(),
      ...this.getRelatedProviders().values(),
    ];
  }

  private linkRelatedProviders() {
    const providers = this.getRelatedProviders();

    providers.forEach(token => {
      const provider = this.container.getProvider(token, this.target);

      this.injector.bind(token).toConstantValue(provider);
    });
  }

  private getRelatedProviders() {
    const providerScope = new Set<Token>();

    const find = (type: OneModule | Token | unknown, scope: Type[]) => {
      if (Registry.isToken(type)) {
        providerScope.add(<Token>type);
      } else if (type instanceof OneModule) {
        for (const related of (<OneModule>type).exports) {
          const ref = this.container.hasModule(<Type<OneModule>>related)
            ? this.container.getModule(<Type<OneModule>>related)
            : related;

          find(ref!, [...scope, type.target]);
        }
      } else {
        throw new InvalidExportException(type, scope);
      }
    };

    for (const related of this.imports) {
      find(related, [this.target]);
    }

    return providerScope;
  }

  private async bindProviders() {
    this.linkRelatedProviders();

    for (const provider of this.providers) {
      const token = Registry.getProviderToken(provider);

      const isMulti = (<MultiDepsProvider>provider).multi;
      if (!isMulti && this.container.hasProviderToken(token)) {
        const name = Registry.getProviderName(provider);
        throw new MultiProviderException(name);
      }

      this.container.addProviderToken(token);
      const type = this.getProviderType(provider);
      await this.bind(token, type, provider);
    }
  }

  public replace(toReplace: Dependency, options: any) {
    this.addProvider({
      provide: <any>toReplace,
      ...options,
    });
  }

  private getModule() {
    return this.injector.get<any>(this.target);
  }

  private runOnAppDestroy(instance: any) {
    return instance.onAppDestroy && instance.onAppDestroy();
  }

  private runOnAppInit(instance: any) {
    return instance.onAppInit && instance.onAppInit();
  }

  private runOnModuleDestroy(instance: any) {
    return instance.onModuleDestroy && instance.onModuleDestroy();
  }

  private runOnModuleInit(instance: any) {
    return instance.onModuleInit && instance.onModuleInit();
  }

  private async everyInjectable(
    every: (provider: Type<any>) => Promise<void> | void,
  ) {
    for (const provider of this.providers) {
      if (!Reflector.isInjectable(provider)) continue;

      await every(this.injector.get(<Type<any>>provider));
    }
  }

  public async onModuleInit() {
    const module = this.getModule();

    await this.runOnModuleInit(module);
    await this.factoriesOnModuleInit();
    await this.injectablesOnModuleInit();
  }

  public async onAppInit() {
    const module = this.getModule();

    await this.runOnAppInit(module);
    await this.factoriesOnAppInit();
    await this.injectablesOnAppInit();
  }

  public async onAppDestroy() {
    const module = this.getModule();

    await this.runOnAppDestroy(module);
    await this.factoriesOnAppDestroy();
    await this.injectablesOnAppDestroy();
  }

  private async injectablesOnAppDestroy() {
    await this.everyInjectable(provider => this.runOnAppDestroy(provider));
  }

  private async injectablesOnAppInit() {
    await this.everyInjectable(provider => this.runOnAppInit(provider));
  }

  private async factoriesOnAppDestroy() {
    await this.runProviderSeries(APP_DESTROY);
  }

  private async factoriesOnAppInit() {
    await this.runProviderSeries(APP_INIT);
  }

  private async injectablesOnModuleInit() {
    await this.everyInjectable(provider => this.runOnModuleInit(provider));
  }

  private async factoriesOnModuleInit() {
    await this.runProviderSeries(MODULE_INIT);
  }

  private async runProviderSeries(token: InjectionToken<any>) {
    await runSeries(this.container.getAllProviders(token, this.target));
  }

  public async create() {
    if (this.injector.isBound(this.target)) return;

    await this.bindProviders();

    this.injector.bind(this.target).toSelf();
    await this.onModuleInit();

    this.created.resolve();
  }

  private getProviderType(provider: Provider) {
    const module = this.target.name;

    if (Reflector.isInjectable(provider)) {
      return ProviderTypes.DEFAULT;
    } else if (Registry.isFactoryProvider(provider, module)) {
      return ProviderTypes.FACTORY;
    } else if (Registry.isValueProvider(provider)) {
      return ProviderTypes.VALUE;
    } else if (Registry.isClassProvider(provider, module)) {
      return ProviderTypes.CLASS;
    } else if (Registry.isExistingProvider(provider, module)) {
      return ProviderTypes.EXISTING;
    }

    throw new Error('');
  }

  public getProvider(ref: ModuleImport): Token {
    const token = Registry.getProviderToken(<Provider>ref);
    const provider = <any>(
      this.getProviders().find(provider => provider === token)
    );
    if (!provider) throw new UnknownProviderException(<any>ref, this.target);
    return provider;
  }

  private getDependencies(dependencies: ModuleImport[] = []) {
    return Promise.all(
      dependencies.map(ref => {
        const dependency = Registry.getForwardRef(ref);
        Registry.assertProvider(dependency);

        const provider = this.getProvider(dependency);
        return this.container.getProvider(provider, this.target);
      }),
    );
  }

  private async bindFactoryProvider(
    token: Token,
    provider: FactoryProvider<any>,
  ) {
    const deps = await this.getDependencies(provider.deps);

    return this.injector.bind(token).toFactory(() => {
      if (
        !this.factoryValues.has(token) ||
        provider.scope === Scopes.TRANSIENT
      ) {
        this.factoryValues.set(token, provider.useFactory(...deps));
      }

      return this.factoryValues.get(token);
    });
  }

  private bindProvider(provider: Type, scope?: Scopes) {
    const binding = this.injector.bind(provider).toSelf();

    switch (scope) {
      case Scopes.TRANSIENT:
        return binding.inTransientScope();

      case Scopes.REQUEST:
        return binding.inRequestScope();

      // case Scopes.SINGLETON:
      default:
        return binding.inSingletonScope();
    }
  }

  private bindClassProvider(token: Token, provider: ClassProvider<any>) {
    return this.injector.bind(token).to(provider.useClass);
  }

  private bindValueProvider(token: Token, provider: ValueProvider<any>) {
    return this.injector.bind(token).toConstantValue(provider.useValue);
  }

  private bindExistingProvider(token: Token, provider: ExistingProvider<any>) {
    const existingToken = Registry.getToken(provider.useExisting);
    const existing = this.injector.get(existingToken);
    return this.injector.bind(token).toConstantValue(existing);
  }

  public async bind(token: Token, type: ProviderTypes, provider: Provider) {
    if (type === ProviderTypes.DEFAULT) {
      const scope = Reflector.getProviderScope(<Type>provider);
      const lazyInjects = Registry.getLazyInjects(<Type>provider);

      lazyInjects.forEach(({ lazyInject, forwardRef }) => {
        const token = Registry.getForwardRef(forwardRef);
        lazyInject(this.lazyInject, token);
      });

      this.bindProvider(<Type>provider, scope);
    } else if (type === ProviderTypes.FACTORY) {
      await this.bindFactoryProvider(token, <FactoryProvider<any>>provider);
    } else if (type === ProviderTypes.VALUE) {
      this.bindValueProvider(token, <ValueProvider<any>>provider);
    } else if (type === ProviderTypes.CLASS) {
      this.bindClassProvider(token, <ClassProvider<any>>provider);
    } else if (type === ProviderTypes.EXISTING) {
      this.bindExistingProvider(token, <ExistingProvider<any>>provider);
    }
  }

  public addGlobalProviders() {
    this.injector.bind(Injector).toConstantValue(this.injector);
    this.injector.bind(OneContainer).toConstantValue(this.container);
    this.injector.bind(MODULE_REF.name).toConstantValue(this);

    /*this.injector
      .bind(Injector)
      .toConstantValue(this.injector)
      .whenInjectedInto(<any>this.target);

    this.injector
      .bind(OneContainer)
      .toConstantValue(this.container)
      .whenInjectedInto(<any>this.target);*/
  }
}
