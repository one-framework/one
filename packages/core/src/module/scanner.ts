import { CircularDependencyException } from '../errors';
import { OneContainer } from './container';
import { Reflector } from '../reflector';
import { Metadata } from '../constants';
import { Registry } from '../registry';
import { OneModule } from './module';
import { concat, omit } from '../util';
import { ModuleExport, ModuleImport, Provider, Type } from '../interfaces';

export class Scanner {
  constructor(private readonly container: OneContainer) {}

  public async scan(module: Type<OneModule>) {
    await this.scanForModules(module);
    await this.scanModulesForDependencies();
    this.container.bindGlobalScope();
    await this.createModules();
  }

  /**
   * @TODO: Fix circular imports with forwardRef
   */
  private async createModules() {
    const createModule = async (module: OneModule) => {
      if (!this.container.isModuleCreated(module)) {
        const imports = [...module.imports.values()];

        // Gets all imports where current module is referenced
        // const circularRefs = imports.filter(({ imports }) => imports.has(module));
        // const importsWithoutRefs = omit<OneModule>(imports, circularRefs);

        // Create all circular referenced modules afterwards
        for (const innerModule of imports) {
          await createModule(innerModule);
        }

        await Promise.all([...imports].map(({ created }) => created));
        await module.create();
        this.container.addCreatedModule(module);

        // Create all circular referenced modules afterwards
        /*for (const innerModule of circularRefs) {
          await createModule(innerModule);
        }*/
      }
    };

    const rootModule = this.container.getRootModule();
    await createModule(rootModule);
  }

  /**
   * Scan recursively for imports in a module
   *
   * @param module
   * @param scope
   * @param ctxRegistry
   */
  private async scanForModules(
    module: Type<OneModule>,
    scope: Type<OneModule>[] = [],
    ctxRegistry = new Set<ModuleImport>(),
  ) {
    module = Registry.getForwardRef(module);

    if (!ctxRegistry.has(module)) {
      ctxRegistry.add(module);

      await this.container.addModule(module, scope);

      const modules = !Registry.isDynamicModule(module)
        ? Reflector.getModuleImports(module)
        : [
            ...Reflector.getModuleImports(module.module),
            ...(module.imports || []),
          ];

      for (const innerModule of modules) {
        /*if (ctxRegistry.has(
          Registry.getForwardRef(innerModule)
        )) continue;*/

        const scopedModules = concat(scope, module);
        await this.scanForModules(innerModule, scopedModules, ctxRegistry);
      }
    }
  }

  public async storeImport(
    related: ModuleImport,
    token: string,
    context: string,
  ) {
    if (!related) throw new CircularDependencyException(context);

    await this.container.addImport(Registry.getForwardRef(related), token);
  }

  public async scanModulesForDependencies() {
    const modules = this.container.getModules();

    for (const [token, { target }] of modules) {
      await this.reflectImports(target, token, target.name);
      await this.reflectProviders(target, token);
      this.reflectExports(target, token);
    }
  }

  private async reflectProviders(module: Type<OneModule>, token: string) {
    const providers = this.getDynamicMetadata<Provider>(
      module,
      token,
      Metadata.PROVIDERS,
    );

    for (const provider of providers) {
      await this.container.addProvider(provider, token);
    }
  }

  private getDynamicMetadata<T>(
    module: Type<OneModule>,
    token: string,
    metadataKey: Metadata,
  ): T[] {
    return [
      ...(Reflector.get(metadataKey, module) || []),
      ...this.container.getDynamicMetadataByToken(token, metadataKey),
    ];
  }

  private reflectExports(module: Type<OneModule>, token: string) {
    const exported = this.getDynamicMetadata<ModuleExport>(
      module,
      token,
      Metadata.EXPORTS,
    );

    exported.forEach(exported => {
      this.container.addExported(exported, token);
    });
  }

  private async reflectImports(
    module: Type<OneModule>,
    token: string,
    context: string,
  ) {
    const modules = this.getDynamicMetadata<ModuleImport>(
      module,
      token,
      Metadata.IMPORTS,
    );

    for (const related of modules) {
      await this.storeImport(related, token, context);
    }
  }
}
