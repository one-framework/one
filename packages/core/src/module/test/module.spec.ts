import { OneModule } from '../module';
import { OneContainer } from '../container';
import { Inject, Module, Injectable } from '../../decorators';
import { InjectionToken } from '../injection-token';
import { ProviderTypes, Scopes } from '../../constants';
import { MultiProviderException, UnknownExportException } from '../../errors';
import { Injector, MODULE_REF } from '../../tokens';
import { noop } from '../../util';
import {
  ClassProvider,
  ExistingProvider,
  FactoryProvider,
  ValueProvider,
} from '../../interfaces';
import { Registry } from '../../registry';

describe('OneModule', () => {
  let container: OneContainer;
  let testProvider: FactoryProvider<any>;
  let module: any;

  const TEST = new InjectionToken<void>('TEST');

  @Module()
  class TestModule {}

  @Injectable()
  class TestService {}

  beforeEach(() => {
    container = new OneContainer();
    module = new OneModule(TestModule, [], container);

    testProvider = {
      provide: TEST,
      useFactory: jest.fn(),
    };
  });

  describe('getDependencies', () => {});

  describe('events', () => {
    beforeEach(() => {});

    describe('onModuleInit', () => {
      describe('runOnModuleInit', () => {});
    });

    describe('onModuleDestroy', () => {});

    describe('onAppInit', () => {
      describe('runOnAppInit', () => {});

      describe('factoriesOnAppInit', () => {});

      describe('injectablesOnAppInit', () => {});
    });

    describe('onAppDestroy', () => {});
  });

  describe('bindings', () => {
    const TestSymbol = Symbol.for('TestService');

    describe('bindProviders', () => {
      beforeAll(() => {
        Registry.getProviderToken = jest.fn(() => TEST.name);
      });

      beforeEach(() => {
        module.linkRelatedProviders = noop;
        module.getProviderType = jest.fn(() => ProviderTypes.FACTORY);
        module.bind = jest.fn();
      });

      afterAll(() => {
        (<any>Registry.getProviderToken).mockClear();
      });

      it('bind should be called with right arguments', async () => {
        module.addProvider(testProvider);

        await expect(module.bindProviders()).resolves.not.toThrow();

        expect(module.bind).toHaveBeenCalledWith(
          TEST.name,
          ProviderTypes.FACTORY,
          testProvider,
        );
      });

      it('should throw MultiProviderException if providers with same identifier exists without multi property set', async () => {
        const error = new MultiProviderException(TEST.name.toString());

        module.addProvider(testProvider);
        module.addProvider(testProvider);

        await expect(module.bindProviders()).rejects.toThrowError(error);
      });

      it('should bind providers with same identifier if multi property is set', async () => {
        testProvider.multi = true;

        module.addProvider(testProvider);
        module.addProvider(testProvider);

        await expect(module.bindProviders()).resolves.not.toThrow();
      });
    });

    describe('bindValueProvider', () => {
      it('should bind a string value', () => {
        const useValue = 'Hello World';

        module.bindValueProvider(TestSymbol, { useValue });

        expect(module.injector.get(TestSymbol)).toEqual(useValue);
      });

      it('should bind any value', () => {
        const useValue = undefined;

        module.bindValueProvider(TestSymbol, { useValue });

        expect(module.injector.get(TestSymbol)).toBeUndefined();
      });
    });

    describe('bindExistingProvider', () => {
      let existingProvider: ExistingProvider<TestService>;

      beforeEach(() => {
        module.injector.bind(TestService).toSelf();

        existingProvider = {
          provide: TEST,
          useExisting: TestService,
        };
      });

      it('should be instance of TestService when using token as identifier', () => {
        module.bindExistingProvider(TestSymbol, existingProvider);

        expect(module.injector.get(TestSymbol)).toBeInstanceOf(TestService);
      });

      it('should be instance of TestService when using class as identifier', () => {
        class Test {}

        module.bindExistingProvider(Test, existingProvider);

        expect(module.injector.get(Test)).toBeInstanceOf(TestService);
      });
    });

    describe('bindFactoryProvider', () => {
      const identifier = Symbol.for('FactoryProvider');
      let counter: number;

      beforeEach(() => {
        counter = 0;
        testProvider = {
          provide: TEST,
          useFactory: jest.fn(() => counter++),
        };
      });

      afterEach(() => {
        (<any>testProvider.useFactory).mockClear();
      });

      it('should bind FactoryProvider as singleton', async () => {
        await module.bindFactoryProvider(identifier, testProvider);

        const firstVal = module.injector.get(identifier);
        const secondVal = module.injector.get(identifier);

        expect(testProvider.useFactory).toHaveBeenCalledTimes(1);
        expect(firstVal).toEqual(secondVal);
      });

      it('should bind FactoryProvider as transient', async () => {
        testProvider.scope = Scopes.TRANSIENT;
        await module.bindFactoryProvider(identifier, testProvider);

        const firstVal = module.injector.get(identifier);
        const secondVal = module.injector.get(identifier);

        expect(testProvider.useFactory).toHaveBeenCalledTimes(2);
        expect(firstVal).not.toEqual(secondVal);
      });

      it('should bind FactoryProvider as request', async () => {});
    });
  });

  describe('validateExported', () => {
    @Module()
    class AnotherTestModule {}

    it('should return token if provider exists', () => {
      const spy = jest.spyOn(module, 'providerContainerHasToken');

      module.addProvider(TestService);

      expect(module.validateExported(TestService, TestService)).toBe(
        TestService,
      );
      expect(spy).toHaveReturnedWith(true);
    });

    it('should return token if module import exists', () => {
      const anotherTestModuleRef = new OneModule(
        AnotherTestModule,
        [],
        container,
      );

      module.addImport(anotherTestModuleRef);

      expect(
        module.validateExported(AnotherTestModule, AnotherTestModule),
      ).toBe(AnotherTestModule);
    });

    it('should throw UnknownExportException if export is not part of imports', () => {
      const error = new UnknownExportException(
        module.target.name,
        AnotherTestModule.name,
      );

      expect(() =>
        module.validateExported(AnotherTestModule, AnotherTestModule),
      ).toThrowError(error);
    });
  });

  describe('addGlobalProviders', () => {
    it('should bind global providers to module', () => {
      @Module()
      class AnotherTestModule {
        constructor(
          readonly container: OneContainer,
          readonly injector: Injector,
          @Inject(MODULE_REF)
          readonly moduleRef: OneModule,
        ) {}
      }

      module = new OneModule(AnotherTestModule, [], container);
      module.addGlobalProviders();

      expect(module.injector.get(Injector)).toBe(module.injector);
      expect(module.injector.get(MODULE_REF.name)).toBe(module);
      expect(module.injector.get(OneContainer)).toBe(module.container);

      module.injector.bind(AnotherTestModule).toSelf();

      const anotherTestModule = module.injector.get(AnotherTestModule);
      expect(anotherTestModule.container).toBe(module.container);
      expect(anotherTestModule.injector).toBe(module.injector);
      expect(anotherTestModule.moduleRef).toBe(module);
    });
  });

  describe('getRelatedProviders', () => {
    let catModule: OneModule;

    @Injectable()
    class CatService {}

    @Module()
    class CatModule {}

    beforeEach(() => {
      catModule = new OneModule(CatModule, [TestModule], container);
    });

    it(`should not have providers which aren't exported`, () => {
      catModule.exports.add(CatService);
      catModule.addProvider(TestService);
      module.addImport(catModule);

      const providers = module.getRelatedProviders();

      expect(providers.size).toBe(1);
      expect(providers.values().next().value).toBe(CatService);
      expect(providers).toMatchSnapshot();
    });

    it('should be able to resolve symbols', () => {
      const TEST = Symbol.for('TEST');

      catModule.exports.add(TEST);
      module.addImport(catModule);

      const providers = module.getRelatedProviders();

      expect(providers.size).toBe(1);
      expect(providers.values().next().value).toBe(TEST);
      expect(providers).toMatchSnapshot();
    });

    it('should be able to resolve Symbol', () => {});

    it('should get deep exported imports/providers', () => {});
  });

  describe('getProviderType', () => {
    it('should get correct provider types', () => {
      const FACTORY_TOKEN = new InjectionToken<void>('FACTORY_TOKEN');
      const VALUE_TOKEN = new InjectionToken<string>('VALUE_TOKEN');
      const EXISTING_TOKEN = new InjectionToken<string>('EXISTING_TOKEN');
      const CLASS_TOKEN = new InjectionToken<TestService>('CLASS_TOKEN');

      expect(module.getProviderType(TestService)).toEqual(
        ProviderTypes.DEFAULT,
      );

      const factoryProvider: FactoryProvider<void> = {
        provide: FACTORY_TOKEN,
        useFactory: () => {},
      };

      expect(module.getProviderType(factoryProvider)).toEqual(
        ProviderTypes.FACTORY,
      );

      const valueProvider: ValueProvider<string> = {
        provide: VALUE_TOKEN,
        useValue: '',
      };

      expect(module.getProviderType(valueProvider)).toEqual(
        ProviderTypes.VALUE,
      );

      const existingProvider: ExistingProvider<string> = {
        provide: EXISTING_TOKEN,
        useExisting: VALUE_TOKEN,
      };

      expect(module.getProviderType(existingProvider)).toEqual(
        ProviderTypes.EXISTING,
      );

      const classProvider: ClassProvider<TestService> = {
        provide: CLASS_TOKEN,
        useClass: TestService,
      };

      expect(module.getProviderType(classProvider)).toEqual(
        ProviderTypes.CLASS,
      );
    });
  });
});
