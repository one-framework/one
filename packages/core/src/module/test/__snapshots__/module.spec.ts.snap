// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneModule getRelatedProviders should be able to resolve symbols 1`] = `
Set {
  Symbol(TEST),
}
`;

exports[`OneModule getRelatedProviders should not have providers which aren't exported 1`] = `
Set {
  [Function],
}
`;
